torch/utils/_device.py

```diff
@@ -65,12 +66,36 @@ def __enter__(self):
         global CURRENT_DEVICE
         self.old_device = CURRENT_DEVICE
         CURRENT_DEVICE = self.device
-        return super().__enter__()
+        # We need to put the device at the bottom of the stack
+        # If we set default device within a function mode context
+        # exiting that context mode will pop the device function mode off
+        # of the stack incorrectly
+        cur_stack = []
+        for _ in range(_len_torch_function_stack()):
+            cur_stack.append(_pop_mode())
+
+        _push_mode(self)
+
+        for mode in reversed(cur_stack):
+            _push_mode(mode)
+
 
     def __exit__(self, exc_type, exc_val, exc_tb):
         global CURRENT_DEVICE
         CURRENT_DEVICE = self.old_device
-        return super().__exit__(exc_type, exc_val, exc_tb)
+        cur_stack = []
+        # Invariant: there should only be one DeviceContext on the stack at any time
+        # search the stack for the device context and pop it, restore previous stack
+        # entries popped while searching
+        for _ in range(_len_torch_function_stack()):
```

Comment:
shouldn't we have an invariant that the bottom-most mode **must** be our `DeviceContext`? (So in particular, we can also assert only the last mode that we pop is of that type?)

Reply:
Yeah I'll add this assertion, it was missing in the previous impl!